Step3 리뷰요청 드립니다.

qwer920414-ctrl opened the pull request:
안녕하세요! 메리크리스마스입니다.
3단계 미션 진행해보았습니다.
아직 더 진행해야하는 부분이 있는거 같기도한데, 우선 Session에 집중해서 구현해보았습니다.

고민됬던 부분은 이렇습니다.
- 인터페이스 EnrollmentPolicy를 어떻게 데이터화해서 DB에 넣는지
- Session과 Enrollment Repository를 어떻게 연결할지

변경을 최소화하려고 했는데, 일부 추가된 부분이 있습니다.
특히 jdbc에 데이터를 넣기 위해서 get 메소드가 많이 추가되었는데요.. 뭔가 찝찝했지만, 필요한거 같아서 추가했습니다.
그리고 SessionRepository에 Enrollment도 넣으려다가 따로 구현하고, 서비스를 만들어서 각각의 Repository를 묶었습니다.
요 방향이 맞는건지...확신은 없지만 서비스로 구현해두니 어느정도 계층이 구별되는거 같아보여요.
( 이렇게 domain/repository/service를 구별해서 실무에서도 개발하면 더 재밌을거 같습니다. )

좀 어설픈거 같아서..  확인 부탁드리고, 피드백 주시면 더 깊게 고민해보겠습니다!

qwer920414-ctrl committed:
feat: session, enrollment 테이블 생성

qwer920414-ctrl committed:
feat: session, enrollment 테이블 수정(type -> name)

qwer920414-ctrl committed:
feat: SessionRepository 생성

qwer920414-ctrl committed:
feat: JdbcSessionRepository 구현
- Jdbc 연결을 위한 get 메소드 추가
- CoverImage fileName 필드 추가
- EnrollmentPolicy 필드 저장을 위한 데이터 추가(type, price)
- Test 코드 구현

qwer920414-ctrl committed:
feat: Session 클래스 toString() 추가

qwer920414-ctrl committed:
feat: EnrollmentRepository 생성
- 인터페이스 생성
- 구현 로직 생성
- 테스트코드 구현

qwer920414-ctrl committed:
style: 패키지 정리
- domain/repository/service로 구분

qwer920414-ctrl committed:
feat: SessionService
- SessionService 구현(repository 묶음)

javajigi commented on src/main/java/nextstep/courses/domain/enrollment/EnrollmentPolicyFactory.java:
```diff
@@ -0,0 +1,20 @@
+    FREE,
+    PAID;
+
+    public static EnrollmentPolicy create(String name, Long price) {
```
EnrollmentPolicy 생성을 PolicyType이 담당하기 보다 별도의 EnrollmentPolicyFactory와 같은 객체를 추가하는 것은 어떨까?

javajigi commented on src/main/java/nextstep/courses/service/SessionService.java:
```diff
@@ -0,0 +1,28 @@
+        this.enrollmentRepository = enrollmentRepository;
+    }
+
+    public void enroll(Long sessionId, Long userId, Payment payment) {
```
👍

javajigi commented on src/main/java/nextstep/courses/domain/Session.java:
메서드 이름이 add이면 반환 타입을 void로 예상할 가능성도 있어보임
AI와의 협업을 통해 더 적합한 이름이 있는지 피드백을 받아 적용해 본다.

javajigi reviewed (changes requested):
db 매핑 작업 넘 잘 했네요. 👍 
기존의 도메인 객체를 최대한 유지하면서 적용한 점이 인상적이네요.

굳이 image와 같은 부분에 대한 작업은 하지 않아도 될 수 있는데요.
이 부분이 4단계 미션에 영향을 미치는 부분이 있어서요.
session외에 다른 부분까지 db 매핑한 후에 미션 마무리하고 4단계 진행하면 좋겠습니다.

추가로 domain 패키지에 도메인 클래스가 증가하고 있네요.
domain 패키지를 분리하는 도전도 추가로 진행해 보면 좋겠습니다.

qwer920414-ctrl committed:
refactor: EnrollmentPolicyFactory 생성
- PolicyType의 역할 분리

qwer920414-ctrl committed:
refactor: Enrollment의 add() 메소드명 수정
- Enrollment 반환을 고려해서 enroll()로 수정

qwer920414-ctrl committed:
refactor: 패키지 구조 수정
- session/enrollment로 패키지 생성

qwer920414-ctrl committed:
feat: Session 클래스에 courseId 추가
- Course -> Sessions 연동 추가
- JdbcCourseRepository 로직 추가
- SessionRepository - findByCourseId 추가

qwer920414-ctrl commented:
안녕하세요!
피드백 주신 부분 적용해보았습니다.
CourseRepository에 Sessions를 추가할때는 CourseRepository에서 SessionRepository를 주입하였습니다.
Course와 Session의 관계상 그렇게 적용해보았는데,
이 부분도 각각의 Repository를 구현하고, Service에서 묶는게 더 좋은 구조일까요?

그리고 get메소드에 대해서 질문드리면,
JDBC로 연동해보니, 갑자기 없던 get 메소드들을 추가할 수 밖에 없었는데 JPA에는 get 메소드 없이 매핑할 수 있는 기능이 있는걸까요?
아직 Mybatis만 사용해보고, JPA를 사용해보지 않았는데 그저 쿼리만 안짜도 된다 이렇게만 알고 있다가 이번 과정을 통해보니 JPA가 도메인 객체 설계방식의 마지막 피날레인거 같네요!

피드백 주시면 고민해보고 적용해보겠습니다! 감사합니다.
( 일정이 계속 있어서 매일 매일 작업하지 못하고 있네요ㅠㅠ 그래도 마감의 그날까지 달려보겠습니다! )

javajigi reviewed (approved):
피드백 반영 잘 했네요. 👍 
패키지 분리까지도 굿.
앞으로 작은 크기의 도메인 객체가 많아질텐데요. 일정 수준이 넘으면 패키지 분리 도전해 보면 좋겠습니다.
아래 질문과 관련한 피드백 반영은 다음 단계에서 함께 반영해 볼 것을 추천합니다.

Q1. CourseRepository에 Sessions를 추가할때는 CourseRepository에서 SessionRepository를 주입하였습니다.
Course와 Session의 관계상 그렇게 적용해보았는데,
이 부분도 각각의 Repository를 구현하고, Service에서 묶는게 더 좋은 구조일까요?
A1. 그리 좋은 접근 방식은 아니라 생각합니다.
Service의 역할이 Repository를 통해 도메인 객체를 생성하고, 도메인 객체 조합을 만드는 역할이기 때문에 Service에서 담당하는 것이 역할에 맞는 접근 방식이라 생각해요.

Q2. 그리고 get메소드에 대해서 질문드리면,
JDBC로 연동해보니, 갑자기 없던 get 메소드들을 추가할 수 밖에 없었는데 JPA에는 get 메소드 없이 매핑할 수 있는 기능이 있는걸까요?
A2. JPA는 리플렉션(Reflection)이라는 기술을 사용하여 private 필드에도 직접 접근할 수 있기 때문에 getter가 없어도 가능하기는 합니다.
JPA는 Getter나 Setter 메소드가 없어도 데이터베이스 테이블과 객체의 필드를 직접 매핑할 수 있는 기능을 지원하고 있어요.
